[SPARK-4979][MLLIB] Streaming logisitic regression

freeman-lab opened the pull request:
This adds support for streaming logistic regression with stochastic gradient descent, in the same manner as the existing implementation of streaming linear regression. It is a relatively simple addition because most of the work is already done by the abstract class `StreamingLinearAlgorithm` and existing algorithms and models from MLlib.

The PR includes
- Streaming Logistic Regression algorithm
- Unit tests for accuracy, streaming convergence, and streaming prediction
- An example use

cc @mengxr @tdas

freeman-lab committed:
Set private and protected within mllib

freeman-lab committed:
Add streaming logistic regression example

freeman-lab committed:
Add streaming logistic regression

freeman-lab committed:
Remove stray comment

SparkQA commented:
[Test build #26503 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26503/consoleFull) for   PR 4306 at commit [`d88f717`](https://github.com/apache/spark/commit/d88f717463263c2a297627981201306f7ab3f94e).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26503 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26503/consoleFull) for   PR 4306 at commit [`d88f717`](https://github.com/apache/spark/commit/d88f717463263c2a297627981201306f7ab3f94e).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class StreamingLogisticRegressionWithSGD (`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26503/
Test FAILed.

freeman-lab committed:
Fixed line length

SparkQA commented:
[Test build #26504 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26504/consoleFull) for   PR 4306 at commit [`85320b1`](https://github.com/apache/spark/commit/85320b17d48fff0fbf621f194e7f5510691d94e4).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26504 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26504/consoleFull) for   PR 4306 at commit [`85320b1`](https://github.com/apache/spark/commit/85320b17d48fff0fbf621f194e7f5510691d94e4).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class StreamingLogisticRegressionWithSGD (`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26504/
Test PASSed.

mengxr commented on examples/src/main/scala/org/apache/spark/examples/mllib/StreamingLogisticRegression.scala:
```diff
@@ -0,0 +1,74 @@
+ * To run on your local machine using the two directories `trainingDir` and `testDir`,
+ * with updates every 5 seconds, and 2 features per data point, call:
+ *    $ bin/run-example \
+ *        org.apache.spark.examples.mllib.StreamingLogisticRegression trainingDir testDir 5 2
```
run-example doesn't need the `org.apache.spark.examples.` prefix on job class names.

mengxr commented on examples/src/main/scala/org/apache/spark/examples/mllib/StreamingLogisticRegression.scala:
```diff
@@ -0,0 +1,74 @@
+
+    ssc.start()
+    ssc.awaitTermination()
+
```
Maybe we should call `sc.stop()` too.

freeman-lab replied:
Wouldn't that kill the whole thing before any data is written? Most of the streaming examples that expect the user to generate a stream themselves don't add a stop (e.g. `StatefulNetworkWordCount`), only the ones that actually generate the data as part of the example (e.g. `QueueStream`). Maybe you're thinking we rewrite the example to generate a bunch of files to feed to the algorithm, and then call `ssc.stop()` at the end?

mengxr replied:
You are right. Just checked `ssc` code and it handles `sc.stop`.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/classification/StreamingLogisticRegressionWithSGD.scala:
```diff
@@ -0,0 +1,97 @@
+import org.apache.spark.mllib.regression.StreamingLinearAlgorithm
+
+/**
+ * Train or predict a logistic regression model on streaming data. Training uses
```
Add `:: Experimental ::` so it can show up in the doc.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/classification/StreamingLogisticRegressionWithSGD.scala:
```diff
@@ -0,0 +1,97 @@
+ *
+ */
+@Experimental
+class StreamingLogisticRegressionWithSGD (
```
Make the constructor private and let users use setters.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/classification/StreamingLogisticRegressionWithSGD.scala:
```diff
@@ -0,0 +1,97 @@
+class StreamingLogisticRegressionWithSGD (
+    private var stepSize: Double,
+    private var numIterations: Int,
+    private var miniBatchFraction: Double,
```
Is it useful? In steaming, we are expecting data coming in small batches. Then it might not be useful to sample them.

freeman-lab replied:
Probably not, though if batches become very large and computing resources are limited, I can imagine a user wanting fractional batches to keep up without lag. So maybe ok to leave in as an option so long as the default is 1.0 (which will definitely be the common choice)?

mengxr replied:
Sounds good.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/classification/StreamingLogisticRegressionWithSGD.scala:
```diff
@@ -0,0 +1,97 @@
+    private var miniBatchFraction: Double,
+    private var regParam: Double,
+    private var initialWeights: Vector)
+  extends StreamingLinearAlgorithm[
```
~~~
  extends StreamingLinearAlgorithm[LogisticRegressionModel, LogisticRegressionWithSGD]
  with Serializable {
~~~

mengxr commented:
LGTM except minor inline comments.

freeman-lab committed:
Fix example formatting

freeman-lab committed:
Experimental tag for docs

freeman-lab committed:
Line formatting

freeman-lab committed:
Make private to mllib

freeman-lab commented:
Thanks for the review! I think I dealt with everything, and in a couple places I tweaked the corresponding point in `StreamingLinearRegression` for parity.

SparkQA commented:
[Test build #26550 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26550/consoleFull) for   PR 4306 at commit [`275f8bd`](https://github.com/apache/spark/commit/275f8bd02254e66ae894459fda25f61c2092db8e).
 * This patch merges cleanly.

SparkQA commented:
**[Test build #26550 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26550/consoleFull)**     for PR 4306 at commit [`275f8bd`](https://github.com/apache/spark/commit/275f8bd02254e66ae894459fda25f61c2092db8e)     after a configured wait of `120m`.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26550/
Test FAILed.

mengxr commented:
test this please

SparkQA commented:
[Test build #26567 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26567/consoleFull) for   PR 4306 at commit [`275f8bd`](https://github.com/apache/spark/commit/275f8bd02254e66ae894459fda25f61c2092db8e).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26567 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26567/consoleFull) for   PR 4306 at commit [`275f8bd`](https://github.com/apache/spark/commit/275f8bd02254e66ae894459fda25f61c2092db8e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26567/
Test FAILed.

freeman-lab committed:
Merge remote-tracking branch 'upstream/master' into streaming-logisitic-regression

freeman-lab commented:
@mengxr the test failure is real, I think there's a conflict due to this recent change to master (https://github.com/apache/spark/commit/b1aa8fe988301b924048039529234278aeb0298a), just merged into my local branch so I can sort it out...

freeman-lab commented:
Ok, the problem is that we've been setting `initialWeights` to `null` by default and then checking that weights are defined before starting training (because we don't know the number of features ahead of time). But that now yields a null pointer exception because the default constructor for `LogisticRegressionModel` checks the size of the weights.

freeman-lab commented:
Should we just set `initialWeights=Vectors.dense(0.0)` by default?

mengxr commented:
Could we try to set model type to `Option[Model]` and check whether it is set in the code?

freeman-lab committed:
Use Option on model

- Forces user to check that weights are initialized, but prevents
needing to initialize weights with null
- Fixes bug due to recent change in LogisitcRegressionModel

freeman-lab commented:
Nice idea! Just took a stab, see what you think. Was reproducing the test failure locally, and it is now fixed with this change.

SparkQA commented:
[Test build #26599 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26599/consoleFull) for   PR 4306 at commit [`5c2c70b`](https://github.com/apache/spark/commit/5c2c70b9eb78a2496a96f8338b7796c76fbf58cd).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26599 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26599/consoleFull) for   PR 4306 at commit [`5c2c70b`](https://github.com/apache/spark/commit/5c2c70b9eb78a2496a96f8338b7796c76fbf58cd).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26599/
Test PASSed.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingLinearAlgorithm.scala:
```diff
@@ -58,14 +58,14 @@ abstract class StreamingLinearAlgorithm[
 
   /** The model to be updated and used for prediction. */
-  protected var model: M
+  protected var model: Option[M] = null
```
This is not the recommended way of using `Option`. I'm going to merge this and send a PR to update it.

freeman-lab replied:
Ah right, `null` -> `None`, and then use `Some`, I see your changes, thanks for fixing that!

mengxr commented:
Merged into master. Thanks!!
